src/Parallel/NodeLock.hpp

```diff
@@ -71,3 +71,31 @@ constexpr inline void unlock(
     const gsl::not_null<NoSuchType*> /*unused*/) noexcept {}
 /// \endcond
 }  // namespace Parallel
+
```

Comment:
Please drop the serialization of the NodeLock. There's a good reason Charm++ doesn't offer it:
```cpp
src/conv-core/converse.h:typedef int CmiNodeLock;
```
it's just an `int` or some other weird type depending on various `ifdef`s. Furthermore, it makes absolutely no sense to send a lock since locks are tied to hardware. The weird code inside the `pup` function also makes it clear that the lock should never be serialized. Additionally, the lock is never destroyed/freed, which makes this code even more difficult to use.

Reply:
Right, so I agree the serialization of the node lock is weird, and my attempt at an implementation can probably be improved, but hear me out -- we _do_ need to serialize objects that contain the NodeLock, because even the groups and node groups get serialized when checkpointing. And when those objects get unpacked, we need to set up the node lock to a usable state.

So, an alternative that would be reasonable (and I started trying before I did this) is to just move any of the node-lock manipulation into the serialization for each separate structure that contains them. It's probably not many, just the `AlgorithmImpl` and the `DataBox`. But, I'm not sure how to approach the serialization there any differently -- it seems as though we should definitely create a new lock when unpacking, and we should probably not free it when packing because (if I understand correctly) it is incorrect to destroy packed objects, instead letting any free occur on destruction.

So, my reasoning was that if we have the same strategy for all cases where we pup something that has the node lock then it may as well be in a centralized location for pupping the node lock, but if it's a preferable design to move it to the objects that contain node locks, I'm fine with that.
Is this misguided, what do you think?

Reply:
I had forgotten we store locks in the DataBox, which complicates things. My concern is actually primarily because of the way Charm++ implements the locks using type aliases. The reason being is that it'll be nearly impossible to avoid adding weird PUP overloads that could cause compiler errors on some architectures. Since we already wrap the locking, unlocking, creation, and destruction I think the best way forward would be to do https://github.com/sxs-collaboration/spectre/issues/998 and write a `Parallel::NodeLock` class that wraps the Charm++ node lock and thus is strongly typed eliminating my concerns. While serializing it is still weird, it then starts to become more reasonable. We can also make it so that the default constructor acquires the lock and the destructor frees the lock. E.g.
```cpp
namespace Parallel {
class NodeLock {
  NodeLock() noexcept {
    lock_ = CmiCreateLock();
  }

  NodeLock(const NodeLock&) = delete;
  NodeLock& operator=(const NodeLock&) = delete;
  NodeLock(NodeLock&&) = default;
  NodeLock& operator=(NodeLock&&) = default;

  ~NodeLock() noexcept {
    CmiDestroyLock(lock_);
  }

  void lock() noexcept {
    CmiLock(lock_);
  }

  bool try_lock() noexcept {
    return CmiTryLock(lock) == 0;
  }

  void unlock() noexcept {
    CmiUnlock(lock_);
  }

  void pup(PUP::er& p) noexcept {}

 private:
  CmiNodeLock lock_{};  // NOLINT
};
}  // namespace Parallel
```

Do you think that would solve the problem?

Reply:
Ah, yeah gotcha -- I agree, then, that the best thing to do is to just add the wrapper class. I especially like that this will do the 'right thing' of always destroying the lock on class destruction, analogous to good smart pointer practice.
I'll add the class as you suggest and replace existing node lock usage with the wrapper in a separate PR (on which this one will then depend) and switch the pup function here to act on the wrapper.
Thanks, @nilsdeppe !

Reply:
That sounds great. Thank you!

I think you only need the member `pup` function in the `NodeLock` class, since that's what we do with all our other classes. There are templates that act on classes with `pup` member functions. Or maybe I'm misunderstanding what you mean by switching the pup function to act on the wrapper.

Reply:
Oh, I think the `move` semantics should do a `CmiCreateLock` and `CmiDestroyLock` since it's not clear that the underlying locks can be moved in any reasonable way, but I think moving a lock in general is well-defined